builder/api/src/main/java/io/helidon/builder/api/ProvidedUtil.java

```diff
@@ -0,0 +1,192 @@
+        Map<String, ConfiguredProvider<T>> unusedProvidersByType = new LinkedHashMap<>();
+        serviceLoader.forEach(it -> {
+            allProvidersByType.put(it.configKey(), it);
+            unusedProvidersByType.put(it.configKey(), it);
+        });
+
+        List<T> result = new ArrayList<>();
+
+        // first add all configured
+        for (ConfiguredService service : configuredServices) {
+            ConfiguredProvider<T> provider = allProvidersByType.get(service.type());
+            if (provider == null) {
+                throw new ConfigException("Unknown provider configured. Expecting a provider with type \"" + service.type()
+                                                  + "\", but only the following providers are supported: "
+                                                  + allProvidersByType.keySet() + ", "
+                                                  + "provider interface: " + providerType.getName()
+                                                  + ", configured service: " + configType.getName());
+            }
+            unusedProvidersByType.remove(service.type());
+            if (service.enabled()) {
+                result.add(provider.create(service.serviceConfig(), service.name()));
+            }
+        }
+
+        // then (if desired) add the rest
+        if (allFromServiceLoader) {
+            unusedProvidersByType.values()
+                    .forEach(it -> result.add(it.create(Config.empty(), it.configKey())));
+        }
+
+        return result;
+    }
+
+    private static ConfiguredService configuredService(Config serviceConfig, boolean isList) {
+        if (isList) {
+            // order is significant
+            String type = serviceConfig.get("type").asString().orElse(null);
+            String name = type;
+            boolean enabled = serviceConfig.get("enabled").asBoolean().orElse(true);
+
+            Config usedConfig = serviceConfig;
+            if (type == null) {
+                // nested approach (we are on the first node of the list, we need to go deeper)
+                List<Config> configs = serviceConfig.asNodeList().orElseGet(List::of);
+                if (configs.size() != 1) {
+                    throw new ConfigException(
+                            "Service provider configuration defined as a list must have a single node that is the type, "
+                                    + "with children containing the provider configuration. Failed on: " + serviceConfig.key());
+                }
+                usedConfig = configs.get(0);
+                name = usedConfig.name();
+                type = usedConfig.get("type").asString().orElse(name);
+                enabled = usedConfig.get("enabled").asBoolean().orElse(enabled);
+            }
+            return new ConfiguredService(usedConfig, type, name, enabled);
+        }
+        // just collect each node, order will be determined by weight
```

Comment:
an interesting option would be to make weight also configurable.

Reply:
That is not really needed - if you want to order it explicitly, you can use a list in configuration, the order will be honored.
If you use an object in configuration, the order will be based on weight of the service provider.
I think that mix and matching these two approaches would not be benefitial

Reply:
My point is that anything in configuration (externalized to the app) would trump a code-based weight assigned.

Reply:
Code based weight is assigned on the service provider level. Configuration is one level higher. This would be even hard to implement, as the information about weight from configuraiton is not available at the point when you resolve service loader (and you may have a single service provider configured more than once, where the weight would be even weirder)